### DISCUSSION TITLE
x-for templates don't refresh or re-render a table correctly

### DISCUSSION THREAD

**Discussion by cartbeforehorse** [alpinejs/alpine#2523] [7. Bugs (archived)]
This post originally started as a question on Stack Overflow, but as my investigation progressed on, it started to feel more and more like a bug in the Alpine engine, hence why I'm now copying it here.

The problem manifests itself in my Laravel project, behind some complicated JavaScript and other integrated functionality.  I've tried as much as possible to isolate the issue to make it comprehensible, but it's not super-easy!  Therefore, the code below is extracted but cut down significantly from my actual application, and this means that become more like pseudo-code than something executable.  Please forgive the odd typo.

Anyway, on with the issue:

I'm using Livewire to synch data between my PHP classes and the AlpineJS front-end.  The two variables that are relevant in the PHP class are:

```lang-php
public $colOrder;        // users are able to "re-order" columns on their table-view.  This preference is saved into their profile and stored in this variable as a 1D array of the column-IDs
public $datasourceData;  // contains a 2D data that is pulled from a database with: Model->get()->toArray(); [0 => ['col1'=>'data1,1', 'col2'=>'data1,2'], 1 => ['col1'=>'data2,1', 'col2'=>'data2,2']];
```

These arrays are then entangled with Alpine variables, and the template is generated from those arrays of data as follows.  Ostensibly, this template works fine:

```lang-html
<div x-data="{
    eColOrder: @entangle('colOrder').defer,
    eData:     @entangle('datasourceData').defer
}">
<table class="table" x-cloak>
  <thead>
	<tr>
	  <template x-for="(col, ix) in eColOrder" :key="'th-'+ix">
		<th x-text="col"></th>
	  </template>
	</tr>
  </thead>
  <tbody>
	<template x-if="eData.length==0">
	  <tr>
		<td :colspan="eColOrder.length" style="padding: 1em">No data found</td>
	  </tr>
	</template>
	<template x-if="eData.length>0">
	  <template x-for="(rec, ix) in eData" :key="'row-'+ix">
		<tr>
		  <td class="action"></td>
		  <template x-for="(col, pos) in eColOrder" :key="'td-'+ix+'-'+pos">
			<td x-text="rec[col]"></td> <!-- I also tried `eData[ix][col]`, but it produced errors in the browser console, even though the on-screen display was fine -->
		  </template>
		</tr>
	  </template>
	</template>
  </tbody>
</table>
```

In this screen-shot, you can see that the user's search (in the top-row) has produced a tabulated grid of data below.  Happy days.

[![enter image description here][1]][1]

The problem arises when the user re-submits a different search.  They do so by updating the search fields, and pressing the "search" button again.  This re-submits the search (through a Livewire JSON call), refreshing the `$datasourceData` array with new data, entangling itself with the `eData` variable in Alpine, and producing the following result:

[![enter image description here][2]][2]

What appears to be happening, is that the results of the new search are correctly pulled through.  But for whatever reason, Alpine hasn't cleared the screen from the last set of search-results.  It's interesting that only the data-level of the HTML table is corrupted (that is to say, the `<td>` cells).  Note that the `<th>` cells have (correctly) not been duplicated above the right-hand-half of the new table.

I've debugged and checked that the data returned from the Eloquent models is correct, and that the structure of the data in the entangled JavaScript variable `eData` is also correct.  This problem is not data-related, it's the rendering that's at fault.

My gut-feel is that this is an Alpine bug, but I haven't been able to prove it yet.

---

My problem stops there.  However, in an attempt to replicate the issue and narrow down the cause of the issue, what I've done is to create a simplified Livewire/Blade/Alpine page.  Strictly speaking, I wasn't able to replicate the problem there either *directly*, but I did (accidentally) manage to replicate a similar output when I entered a deliberate "bug" into my code.

Take the following PHP/Livewire component:

```lang-php
<?php

namespace App\Business\Tbd;

use Livewire\Component;

class StartLw extends Component
{
    public  array $data = [];
    public  array $headings = [];
    public  int $count = 0;

    public function mount() {
        for ($i=1; $i <= 6; $i++) {
            $this->headings[] = "col{$i}";
        }
        $this->data = [];
    }

    public function formSubmit() {
        $src = 1;
        $this->data = [];
        for ($i=0; $i < 10; $i++) {
            $this->data[$i] = [];
            for ($y=1; $y <= 6; $y++) {
                $this->data[$i]["col{$y}"] = "source {$src} ({$i},{$y})";
            }
        }
        $this->count++;
    }

    public function relatedToButSeparateFromForm() {
        $src = 2;
        $this->data = [];
        for ($i=0; $i < 4; $i++) {
            $this->data[$i] = [];
            for ($y=1; $y <= 6; $y++) {
                $this->data[$i]["col{$y}"] = "source {$src} ({$i},{$y})";
            }
        }
        $this->count++;
    }

    public function render()
    {
        return view('components.tbd.lw-start-lw')
            ->layout('layouts.tbd.lw');
    }
}
```

And this cut-down HTML to render the page:

```lang-html
<div class="container" x-data="{
    eData: @entangle('data').defer,
    eHeadings: @entangle('headings').defer
}">

<div class="row">
    <div class="col"><p>{{ $count }}</p></div>
</div>

<div class="row">
    <div class="col">
        <form method="post" wire:submit.prevent="formSubmit">
            <p>
            <button type="submit">Load data source 1</button>&nbsp;
            <button type="button" wire:click="relatedToButSeparateFromForm">Load data source 2</button>
            </p>
        </form>
    </div>
</div>

<div class="row">
    <div class="col">
        <table>
            <thead>
                <tr>
                <template x-for="hd in eHeadings">
                    <th x-text="hd" style="padding: 0.5em; background-color:rgb(220,220,230); border: 1px solid rgb(210,210,230)"></th>
                </template>
                </tr>
            </thead>
            <tbody>
                <template x-for="(row, ix) in eData" :key="ix">
                    <tr>
                    <template x-for="(col, pos) in eHeadings" :key="'td-'+ix+'-'+pos">
                        <td x-text="row[col]" :class="id" style="padding: 0.5em; background-color:rgb(240,240,255); border: 1px solid rgb(210,210,230)"></td>
                    </template>
                    </tr>
                </template>
            </tbody>
        </table>
    </div>
</div>
</div>
```

Note the deliberate error!  On the `<td>` element, `:class="id"` should really say `:class="col"`.  Now if I take out the error, the page works as I would expect it to.  But with the error re-introduced into the code (together with a bunch of error messages in the browser console saying: `Uncaught ReferenceError: id is not defined`), after toggling the two buttons back and forth a bit, I get this:

[![enter image description here][3]][3]

As I think you'll agree, that picture is spookily reminiscent of the situation I get in my real world application (except that in the real-world-app, I don't end up with any errors in the browser's console).

This leads me to the strong belief that there is a silent bug being triggered somewhere in the Alpine engine which triggers the same net result.  Is anyone here able to validate my suspicions?

  [1]: https://i.stack.imgur.com/68ifq.png
  [2]: https://i.stack.imgur.com/i6eS7.png
  [3]: https://i.stack.imgur.com/yzuV5.png

**Comment by SimoTod** [comment]
Were you able to replicate the bug without livewire? I suspect the bug sits in the livewire bit which codes the integration with Alpine (most of the logic to integrate the 2 libraries is in the livewire codebase, including js code).

**Reply by cartbeforehorse** [reply]
Interesting comment.  And no - I can't replicate only using Alpine.  I assumed that Livewire's responsibility ended at the point of passing (or mapping) the data within the PHP variable, to a JavaScript variable.  I didn't realise that Livewire may also be responsible for the rendering of the page.

Should I therefore go over to the Livewire repository and log the same issue there?

**Reply by SimoTod** [reply]
It's a bit of a blurred border. Alpine has some js bits that are there for livewire but also livewire has some js files to integrate with Alpine. 

For instance, `@entangle` and `$wire` are defined in the livewire codebase. The dom diff/update is run by livewire as well and I think it stops the Alpine mutation observer which is probably the reason why the old loop entry don't get cleaned up properly.

have you tried to use wire:ignore on the Alpine generated table? There's no reason for Livewire to try to re process that bit of html

**Reply by cartbeforehorse** [reply]
@SimoTod   Spot on!

Placing `wire:ignore` on the parent `<div>` solves it.  So that confirms that it's a Livewire issue (or at least an issue that lies half-way between Livewire/Alpine).  Amazing what a little bit of deeper understanding can bring!  Thanks for the help.

Still, would you go so far as to say that this is a bug?  Or just a glitch of how the technology works?  In other words, is it worth posting the problem over on the Livewire pages and making them aware of it for future improvement?

**Reply by SimoTod** [reply]
It does look like a bug but there may be techical reasons why it behaves that way (currently it uses morphdom which is a third party library but i believe it will eventually switch to Alpine.morph that will give more control to Caleb). I don't know the livewire internal so I have no idea, you can try to ask on that repo.
